pulsar-client/src/main/java/org/apache/pulsar/client/impl/PulsarServiceNameResolver.java

```diff
@@ -96,15 +130,125 @@ public void updateServiceUrl(String serviceUrl) throws InvalidServiceURL {
+            return oldStatus;
+        });
+
+        hostAvailabilityMap.forEach((__, endpointStatus) -> {
+            if (!endpointStatus.isAvailable()) {
+                computeEndpointStatus(false, endpointStatus);
+                if (!availableHostsChanged.get() && endpointStatus.isAvailable()) {
+                    availableHostsChanged.set(true);
+                }
+            }
+        });
+
+        if (availableHostsChanged.get()) {
+            availableAddressList = hostAvailabilityMap.entrySet()
+                    .stream()
+                    .filter(entry -> entry.getValue().isAvailable() && allAddressSet.contains(entry.getKey()))
+                    .map(Map.Entry::getKey)
+                    .collect(Collectors.toList());
+            log.info("service name resolver available hosts changed, current available hosts: {}",
+                    availableAddressList);
+        }
+    }
+
+    /**
+     * Create an {@link EndpointStatus} for the given address.
+     * @param isAvailable the availability status of the endpoint
+     * @param inetSocketAddress the address of the endpoint
+     * @return a new {@link EndpointStatus} instance
+     */
+    private EndpointStatus createEndpointStatus(boolean isAvailable, InetSocketAddress inetSocketAddress) {
+        Backoff backoff = new BackoffBuilder()
+                .setInitialTime(serviceUrlQuarantineInitDurationMs, TimeUnit.MILLISECONDS)
+                .setMax(serviceUrlQuarantineMaxDurationMs, TimeUnit.MILLISECONDS)
+                .create();
+        EndpointStatus endpointStatus =
+                new EndpointStatus(inetSocketAddress, backoff, System.currentTimeMillis(), 0,
+                        isAvailable);
+        if (!isAvailable) {
+            computeEndpointStatus(false, endpointStatus);
+        }
+        return endpointStatus;
+    }
+
+    /**
+     * Compute the endpoint status based on the new availability status.
+     * @param newIsAvailable the new availability status of the endpoint
+     * @param status the current status of the endpoint
+     */
+    private void computeEndpointStatus(boolean newIsAvailable, EndpointStatus status) {
+        if (!newIsAvailable) {
+            if (!status.isAvailable()) {
+                // from unavailable to unavailable, check if we need to try to recover
+                long elapsedTimeMsSinceLast = System.currentTimeMillis() - status.getLastUpdateTimeStampMs();
+                boolean needTryRecover = elapsedTimeMsSinceLast >= status.getNextDelayMsToRecover();
+                if (needTryRecover) {
+                    log.info("service name resolver try to recover host {} after {}", status.getSocketAddress(),
+                            Duration.ofMillis(elapsedTimeMsSinceLast));
+                    status.setAvailable(true);
+                    status.setLastUpdateTimeStampMs(System.currentTimeMillis());
+                    status.setNextDelayMsToRecover(status.getQuarantineBackoff().next());
```

Comment:
```suggestion
                    status.setNextDelayMsToRecover(0);
                    status.getQuarantineBackoff().reset();
```
If set to `true`, you should reset the delay and backoff.

Reply:
Here is the recovery action performed when the quarantine time expired. If the next call still fails, the quarantine time of the address needs to continue to increase exponentially. For example, suppose host1 fails to call for the first time, available is set to false, and the initial isolation time is 1 minute. After one minute, the quarantine time expires and the address is recovery. If the first call after recovery fails again, the quarantine time should increase exponentially to 2 minutes, and so on.
**Only when the first call after the address is recovery is successful, `quarantineBackoff` will be reset to the initial state.**